scripts/copyrighter.py

```diff
@@ -0,0 +1,156 @@
+#!/usr/bin/env python
+
+# Copyright © 2012-2023 Forschungszentrum Jülich GmbH
+# SPDX-License-Identifier: LGPL-3.0-or-later
+import argparse
+import datetime
+import re
+from pathlib import Path
+
+
+def collect_files(
+    file_name_patterns: list[str],
+    exclude_directories: list[str],
+    start_dir: Path,
+):
+    files = []
+    for pattern in file_name_patterns:
+        for p in start_dir.rglob(f"**/{pattern}"):
+            if exclude_directories is None or not any(
+                [str(p).startswith(dir) for dir in exclude_directories]
+            ):
+                files.append(p)
+
+    return files
+
+
+class Handler:
+    def __init__(
+        self,
+        year: int,
+        extensions: list[str],
+        comment_template: str,
+        account_for_shebang: bool,
+    ):
+        self.comment = [
+            comment_template.format(line)
+            for line in [
+                "Copyright © 2012-{} Forschungszentrum Jülich GmbH".format(
+                    year
+                ),
+                "SPDX-License-Identifier: LGPL-3.0-or-later",
+            ]
+        ]
+        self.extensions = extensions
+        self.account_for_shebang = account_for_shebang
+        self.header_pattern = re.compile(
+            "Copyright © 2012-\\d+ Forschungszentrum Jülich GmbH"
+        )
+
+    def glob_pattern(self) -> list[str]:
+        return [f"*{e}" for e in self.extensions]
+
+    def _write_copyright_header(
+        self, file_contents: list[str], has_shebang: bool
+    ) -> list[str]:
+        new_content: list[str] = []
+        if has_shebang:
+            new_content += file_contents[0:2]
+            new_content += self.comment
+            if len(file_contents) >= 2:
+                new_content += file_contents[2:]
+        else:
+            new_content = self.comment.copy()
+            new_content += file_contents
+
+        return new_content
+
+    def _update_copyright_header(
+        self, file_contents: list[str], has_shebang: bool
+    ) -> list[str]:
+        new_content: list[str] = []
+        if has_shebang:
+            new_content += file_contents[0:2]
+            new_content.append(self.comment[0])
+            if len(file_contents) >= 4:
+                new_content += file_contents[3:]
+        else:
+            new_content.append(self.comment[0])
+            new_content += file_contents[1:]
+        return new_content
+
+    def _has_copyright_header(
+        self, file_contents: list[str], has_shebang: bool
+    ) -> bool:
+        if has_shebang:
+            return (
+                len(file_contents) >= 4
+                and self.header_pattern.search(file_contents[2]) != None
```

Comment:
When only checking for `file_contents[2]` the following case is not handled properly:
```
#!/usr/bin/env python
# Copyright header
```
See annotated files. 
Won't it be better to check the first non empty line, if it is a copy right header?